Add layout and blocksize arguments to Tensor.to_sparse method

pearu opened the pull request:
This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89912
* __->__ #89502

pearu committed:
Add layout and blocksize arguments to Tensor.to_sparse method

[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

amjames commented on aten/src/ATen/SparseCsrTensorUtils.h:
```diff
@@ -287,5 +294,23 @@ inline Layout flip_compressed_layout(Layout layout) {
 }
 
+inline DimVector getBlockSize(Tensor const& self) {
+  int64_t n_batch;
+  switch (self.layout()) {
+    case kSparseBsr:
+      n_batch = self.crow_indices().dim() - 1;
+      break;
+    case kSparseBsc:
+      n_batch = self.ccol_indices().dim() - 1;
+      break;
+    default:
+      return {};
+  }
```
```suggestion
  int64_t n_batch = numBatchDimensions(self);
```

amjames commented on test/test_sparse.py:
```diff
@@ -4100,6 +4109,82 @@ def test_generate_simple_inputs(self):
+                    t.to_sparse(layout=to_layout, blocksize=blocksize)
+                continue
+            elif from_layout in {torch.sparse_csr, torch.sparse_csc} and to_layout is torch.sparse_coo and is_batch:
+                with self.assertRaisesRegex(RuntimeError, "crow_indices is supposed to be a vector"):
```
Could this not-implemented case be caught further up the line and provide a more informative error?

pearu replied:
This error is triggered when trying to convert the indices of a CSR/CSC tensor with batch dimensions to COO indices using `_convert_indices_from_csr_to_coo` function.

I have updated the exception message with the dimensionality information of the `crow_indices`.

We should implement the batch dimensions support in `_convert_indices_from_csr_to_coo` function as it will be needed in many conversion cases. It's a good follow-up task.

amjames commented on test/test_sparse_csr.py:
```diff
@@ -2846,7 +2846,7 @@ def _to_from_layout(layout_a, layout_b, a):
                     for block_layout in block_layouts:
-                        with self.assertRaisesRegex(RuntimeError, "blocksize does not match the blocksize"):
+                        with self.assertRaisesRegex(RuntimeError, "conversion from.*to.*is not implemented"):
                             self._convert_to_layout(b, block_layout, blocksize=3)
```
Related to my comment above about the early exit and silently incorrect behavior. The `self._convert_to_layout` utility use the `to_sparse_xyz` api directly so it bypassed that early exit and triggers the expected error.

This test should probably be re-written to use `to_sparese(layout, blocksize)` that utility is made obsolete by this PR.

pearu replied:
Agreed. I have moved the validation checks from `_convert_to_layout` to `test_sparse_to` test method and eliminated `_convert_to_layout` as well as updated other conversion tests.

amjames reviewed (changes requested):
See comments. One major item, the rest are nits/suggestions

pearu commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -1306,7 +1307,7 @@ Tensor sparse_compressed_to_sparse_csr(const Tensor& self) {
 Tensor coo_to_sparse_csr(const Tensor& self) {
   TORCH_CHECK(
-      self.dim() == 2,
+      self.sparse_dim() == 2,
       "Only 2D tensors can be converted to the SparseCsr layout but got shape: ",
       self.sizes());
```
Applied. Thanks!

pearu commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -1326,7 +1327,7 @@ Tensor coo_to_sparse_csr(const Tensor& self) {
   TORCH_CHECK(
-      self.dim() == 2,
+      self.sparse_dim() == 2,
       "Only 2D tensors can be converted to the SparseCsc layout but got shape: ",
```
Applied.

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

amjames reviewed (approved):
Looks good!

cpuhrsch commented:
@pytorchbot merge

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

pearu committed:
Update on "Add layout and blocksize arguments to Tensor.to_sparse method"

This PR extends the `Tensor.to_sparse()` method to `Tensor.to_sparse(layout=None, blocksize=None)` in a BC manner (`layout=None` means `layout=torch.sparse_coo`).

In addition, the PR adds support for the following conversions:
- non-hybrid/hybrid COO tensor to CSR or CSC or a COO tensor
- short, bool, byte, char, bfloat16, int, long, half CSR tensor to a BSR tensor

and fixes the following conversions:
- hybrid COO to COO tensor
- non-batch/batch hybrid BSR to BSR or BSC tensor




[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
Thanks! It's nice to have a unified conversion function.

nikitaved commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -1341,15 +1342,11 @@ Tensor coo_to_sparse_csc(const Tensor& self) {
-  AT_ERROR(
-      "Conversion from ", self.layout(), " to SparseBsc is currently not supported.");
-  return self;
+  return self.to_sparse_bsr(blocksize).to_sparse_bsc(blocksize);
```
@pearu , could you please tell why there is a double conversion here? `to_sparse_bsc` does seem to have a COO dispatch...

pearu replied:
@nikitaved `self` here is a COO tensor and to convert it to BSC, we'll use an intermediate BSR tensor because we don't have direct COO->BSC implemented yet.

The COO dispatch in `to_sparse_bsc` is exactly this function here.

nikitaved replied:
Got it, thanks! Yet another thing that could be improved :)
